src/Confix.Tool/src/Confix.Tool/Variables/VariableResolver.cs

```diff
@@ -46,30 +49,70 @@ public async Task<IEnumerable<VariablePath>> ListVariables(string providerName,
 
     public async Task<JsonNode> ResolveVariable(VariablePath key, CancellationToken cancellationToken)
     {
+        App.Log.ResolvingVariable(key);
         var configuration = GetProviderConfiguration(key.ProviderName);
         await using var provider = _variableProviderFactory.CreateProvider(configuration);
-        return await provider.ResolveAsync(key.Path, cancellationToken);
+        var resolvedValue = await provider.ResolveAsync(key.Path, cancellationToken);
+
+        if (resolvedValue.IsVariable(out VariablePath? variablePath))
+        {
+            return await ResolveVariable(variablePath.Value, cancellationToken);
+        }
+        return resolvedValue;
     }
 
     public async Task<IReadOnlyDictionary<VariablePath, JsonNode>> ResolveVariables(
         IReadOnlyList<VariablePath> keys,
         CancellationToken cancellationToken)
     {
-        Dictionary<VariablePath, JsonNode> resolvedVariables = new();
+        List<KeyValuePair<VariablePath, JsonNode>> resolvedVariables = new(keys.Count);
 
         foreach (IGrouping<string, VariablePath> group in keys.GroupBy(k => k.ProviderName))
         {
-            var providerConfiguration = GetProviderConfiguration(group.Key);
-            await using IVariableProvider provider = _variableProviderFactory.CreateProvider(providerConfiguration);
-
-            var resolved = await provider.ResolveManyAsync(
-                group.Select(x => x.Path).Distinct().ToArray(),
+            var providerResults = await ResolveVariables(
+                group.Key,
+                group.Select(k => k.Path).ToHashSet(),
```

Comment:
Instead of converting to hashset and then to array we could use linq `Distinct()` which would reveal the intend better in my opinion.

Reply:
fair